File: leetcode/rust-leetcode/src/437.path-sum-iii.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;
type TypeNode = Option<Rc<RefCell<TreeNode>>>;
impl Solution {
    pub fn path_sum(root: TypeNode, target_sum: i32) -> i32 {
        fn dfs(root: &TypeNode, path: &mut Vec<i64>, target_sum: i64, ans: &mut i32) {
            match(root) {
                Some(node) => {
                    let node = node.borrow();
                    let mut sum = node.val as i64;
                    if sum == target_sum {
                        *ans += 1;
                    }
                    for i in (0..path.len()).rev() {
                        sum += path[i];
                        if sum == target_sum {
                            *ans += 1;
                        }
                    }
                    path.push(node.val as i64);
                    if node.left.is_some() {
                        dfs(&node.left, path, target_sum, ans);
                    }
                    if node.right.is_some() {
                        dfs(&node.right, path, target_sum, ans);
                    }
                    path.pop();
                }
                None => {}
            }
        }
        let mut path = Vec::new();
        let mut ans = 0i32;
        dfs(&root, &mut path, target_sum as i64, &mut ans);
        ans
    }
}

File: leetcode/python/437.path-sum-iii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> int:
        count = 0
        def dfs(node: Optional[TreeNode], path: List[int]):
            if node is None:
                return
            path.append(node.val)
            cumSum = 0
            for nval in path[::-1]:
                cumSum += nval
                if cumSum == targetSum:
                    nonlocal count
                    count += 1
            dfs(node.left, path)
            dfs(node.right, path)
            path.pop()
        
        dfs(root, [])
        return count

""" """

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> int:
        
        d = collections.defaultdict(int)
        d[0] = 1

        def dfs(node: Optional[TreeNode], total: int):
            if node is None:
                return 0
            
            total += node.val
            count = d[total - targetSum]

            d[total] += 1

            count += (dfs(node.left, total) + dfs(node.right, total))

            d[total] -= 1

            return count
        
        
        return dfs(root, 0)

